toolkit/components/uniffi-bindgen-gecko-js/src/templates/Scaffolding.h

```diff
@@ -21,6 +23,21 @@ namespace mozilla::dom {
 
 class GlobalObject;
 
+{%- for object in ci.iter_object_definitions() %}
+class {{ object.nm() }}PointerType : public UniFFIPointerType {
+    public:
+        static {{ object.nm() }}PointerType& getInstance() {
+            static {{ object.nm() }}PointerType instance;
+            return instance;
+        }
+    private:
+        {{ object.nm() }}PointerType() {
+            typeName = u"{{ ci.cpp_namespace() }}{{ object.nm() }}"_ns;
```

Comment:
it will be very painful to check that the arguments are correct on function calls

Arguments in the C++ functions have a type of `RustArcPtr`, which does **not** hold the name of the original object. We have two ways of dealing with this.
1) Have `uniffi` add more information that can help us here, fundementally we will be trying to figure out "which object does this pointer belong to" - we can have uniffi have that by default
2) Create some logic in the `cpp.rs` to resolve a pointer to an object type, very similar to what I do in [the new function for return type](https://github.com/bendk/gecko-dev/pull/18/files#diff-82947effc4e43a401fe272bfbd560e582ad634d0ea84293f16a1389db5816866R40) but have it for arguments, and use an argument position to help us out.

the logic will have to be pretty smart (i.e look at all functions, methods and constructors) but that's not too far from what I do for return types, so we can create some shared logic there

Reply:
Alternatively, we can answer the question of "how possible is it for a JS function to pass a wrong pointer to C++, since the code on both sides is generated" - if the answer is "only when things are very very wrong" then we might be able to get away with just checking that a `RustArcPtr` is in fact, a pointer without verifying if it's the "right" pointer type.

If the answer is "there is a good enough chance it happens and disastrous/undefined behaviour will occur" then we can add the handling I mention in the above comment

Reply:
I want to also point out that regardless, we do want to keep the `UniFFIPointerType` type, because we want to call the right destructor when the object is destructed.

Reply:
sadly I don't have anything meaningful to help you here - at the very least, working out how disastrous it would be is probably important (ie, if it's just that js gets the wrong type, that seems ok)